### ISSUE TITLE
Apple M1, pip requirements unavailable 

### ISSUE THREAD

**Issue by carlosvigil** [jaakkopasanen/AutoEq#283]
Just reference for others. I believe it was the Math..something, or Mat, that wouldn't build. You'll know when Terminal turns red.

**Comment by JaKasb** [comment]
The Apple M1 architecture can't execute the scientific python binaries.
https://stackoverflow.com/a/65095679

However this problem is not related to AutoEq...

**Comment by carlosvigil** [comment]
Great link, thank you. 
Relevant: 
> ...One solution is to install a Python distribution like miniconda and run its pip installer to add Intel packages (the scripts included in this repository require numpy, pandas, and seaborn). There is a delay when one first imports one of these libraries while they are translated. Subsequently, the translated Python libraries appear to run well.
> - MATLAB is used by many scientific tools, including SPM. While Matlab works in translation, it is not yet available natively (and mex files will need to be recompiled).

**Comment by misty-** [comment]
This is the fix, all required python modules in 'requirements.txt' install correctly:

arch -x86_64 bash
cd ~/AutoEQ
python3 -m venv venv
. venv/bin/activate
python -m pip install -U pip
python -m pip install -U -r requirements.txt
python autoeq.py --help

**Comment by misty-** [comment]
Although the I was able to setup AutoEq with the commands in my previous post there is trouble when trying to create an EQ graph which will probably be beyond what I can find a fix for.

I tried to create a custom EQ graph for Sony WF-1000XM4... here is the result:

(venv) python autoeq.py --input_dir="my_data/Custom Sony WF-1000XM4" --output_dir="my_results/Custom Sony WF-1000XM4" --compensation="compensation/harman_in-ear_2019v2_wo_bass.csv" --equalize --parametric_eq --max_filters=5+5 --ten_band_eq --convolution_eq --bass_boost=6 --standardize_input
Illegal instruction: 4
(venv)

It returns with failure message "Illegal instruction: 4".

Not sure how to proceed from here.

Side note: I did setup AutoEq on a Win10 laptop and ran the same pass for EQ graph for WF-1000XM4 and got good results, no failure.

**Comment by misty-** [comment]
Here's the fully working solution to getting AutoEq working on an Arm64 (Apple Silicon) M1 Mac. I tested this 3 times to be sure there were no mistakes.

1. Install Xcode from app store, Xcode Command Line Tools should install,
   if not, open terminal and install with: $ xcode-select --install

2. Check for latest version of Miniforge3 at https://github.com/conda-forge/miniforge#download

   Download the file for arm64 (Apple Silicon).

3. Launch terminal and navigate to directory where you saved "Miniforge3-MacOSX-arm64.sh" file.

   Example: If you saved to your User Downloads directory

   $ cd ~/Downloads

4. Install Miniforge3:

   $ bash Miniforge3-MacOSX-arm64.sh

   You must accept the licensing agreement to continue the installation.

5. After complete installation, you are prompted to initialize Miniforge3.

   Do you wish the installer to initialize Miniforge3
   by running conda init? [yes|no]
   [no] >>> yes

6. Python 3.9 is default for Miniforge installation. AutoEq requires python 3.8.
   Next create a virtual environment and install python 3.8.

   (base) MAC_Name:~ Username$ conda create --name autoeq_venv python=3.8 pillow matplotlib pandas scipy tabulate libsndfile absl-py google-auth google-pasta astunparse flatbuffers gast grpcio h5py keras-preprocessing markdown oauthlib opt-einsum protobuf requests requests-oauthlib tensorboard

   Activating virtual environment.

   (base) MAC_Name:~ Username$ conda activate autoeq_venv

   Terminal prompt changes to:

   (autoeq_venv) MAC_Name:~ Username$

   TIP: To deactivate a conda virtual environment use:

        $ conda deactivate


   NOTE: In the normal AutoEq setup instructions you would create and activate a venv directory within the AutoEq directory.
              With Miniforge3 you will be creating and activating a venv directory within ~/miniforge3/envs directory.

7. Installing dependencies outside of conda-forge

   $ conda install -c bricew soundfile

   $ pip install tensorboard-data-server

   $ conda install -c isuruf/label/tf tensorflow-addons

   Apple has an alpha version of tensorflow and tensorflow-addons at GitHub which works for arm64 (Apple Silicon).
   At the time of putting these instructions together the current alpha release is v0.1alpha3.
   Check for a later version at https://github.com/apple/tensorflow_macos/releases before proceeding.
   Click the small triangle next to the word 'Assets' of the latest release to see the file links.

   Copy the url of both arm64 assets. Example below.

   https://github.com/apple/tensorflow_macos/releases/download/v0.1alpha3/tensorflow_addons_macos-0.1a3-cp38-cp38-macosx_11_0_arm64.whl

   https://github.com/apple/tensorflow_macos/releases/download/v0.1alpha3/tensorflow_macos-0.1a3-cp38-cp38-macosx_11_0_arm64.whl

   Paste the urls into this command:

   pip install --upgrade --force --no-dependencies <url> <url>

   Below is the command after inserting the urls, which at the time of writing can be copied and pasted into terminal to install the latest alpha release.

   $ pip install --upgrade --force --no-dependencies https://github.com/apple/tensorflow_macos/releases/download/v0.1alpha3/tensorflow_addons_macos-0.1a3-cp38-cp38-macosx_11_0_arm64.whl https://github.com/apple/tensorflow_macos/releases/download/v0.1alpha3/tensorflow_macos-0.1a3-cp38-cp38-macosx_11_0_arm64.whl

8. Verifying AutoEq.

   $ cd ~/AutoEq

   $ python autoeq.py --help  <<-- This is only a minimal verification!! AutoEq help may be displayed even if setup is a little wonky.

   It's better to run a full EQ graph pass.

   Here's an example for true verification from https://github.com/jaakkopasanen/AutoEq#equalizing-individual-headphones :

   Equalizing Sennheiser HD 650 and saving results to my_results/HD650:

   $ python autoeq.py --input_dir="measurements/innerfidelity/data/onear/Sennheiser HD 650" --output_dir="my_results/HD650" --compensation="measurements/innerfidelity/resources/innerfidelity_harman_over-ear_2018_wo_bass.csv" --equalize --bass_boost=4 --show_plot --convolution_eq --fs=44100,48000

9. Keeping conda updated.

   $ conda update conda --all

**Comment by misty-** [comment]
TensorFlow v2.5 (non-alpha, yay!) for Arm64 (Apple Silicon) M1 Macs is available.

Updated the instructions... Enjoy!


1. Install Xcode from app store, Xcode Command Line Tools should install,

   if not, open terminal and install with: $ xcode-select --install


2. Check for latest version of Miniforge3 at https://github.com/conda-forge/miniforge#download

   Download the file for arm64 (Apple Silicon).


3. Launch terminal and navigate to directory where you saved "Miniforge3-MacOSX-arm64.sh" file.

   Example: If you saved to your User Downloads directory

   
   $ cd ~/Downloads


4. Install Miniforge3:

   $ bash Miniforge3-MacOSX-arm64.sh

   You must accept the licensing agreement to continue the installation.


5. After complete installation, you are prompted to initialize Miniforge3.

   Do you wish the installer to initialize Miniforge3
   by running conda init? [yes|no]
   [no] >>> yes


6. Python 3.9 is default for Miniforge installation. AutoEq requires python 3.8.

   Next create a virtual environment and install python 3.8.

   (base) MAC_Name:~ Username$ conda create --name autoeq_venv python=3.8 pillow matplotlib pandas scipy tabulate libsndfile

   Activating virtual environment.

   (base) MAC_Name:~ Username$ conda activate autoeq_venv

   Terminal prompt changes to:

   (autoeq_venv) MAC_Name:~ Username$

   TIP: To deactivate a conda virtual environment use:

        $ conda deactivate

   NOTE: In the normal AutoEq setup instructions you would create and activate a venv directory within the AutoEq directory.
         With Miniforge3 you will be creating and activating a venv directory within ~/miniforge3/envs directory.


7. Installing dependencies outside of conda-forge

   $ conda install -c bricew soundfile


8. Installing TensorFlow v2.5

   $ conda install -c apple tensorflow-deps
   
   $ python -m pip install tensorflow-macos

8a. Optional installation of tensorflow-metal PluggableDevice to accelerate training with Metal on Mac GPUs.
  
   $ python -m pip install tensorflow-metal

NOTE: I tried running AutoEq EQ passes with tensorflow-metal installed and with it not installed.
I could not tell for sure, but it seemed that EQ passes with tensorflow-metal installed were marginally faster.
Just know that AutoEq will work either way.


9. Verifying AutoEq.

   $ cd ~/AutoEq

   $ python autoeq.py --help  <<-- This is only a minimal verification!! AutoEq help may be displayed even if setup is a little wonky.

   It's better to run a full EQ graph pass.

   Here's an example for true verification from https://github.com/jaakkopasanen/AutoEq#equalizing-individual-headphones :

   Equalizing Sennheiser HD 650 and saving results to my_results/HD650:

   $ python autoeq.py --input_dir="measurements/innerfidelity/data/onear/Sennheiser HD 650" --output_dir="my_results/HD650" --compensation="measurements/innerfidelity/resources/innerfidelity_harman_over-ear_2018_wo_bass.csv" --equalize --bass_boost=4 --show_plot --convolution_eq --fs=44100,48000


10. Keeping conda updated.

   $ conda update conda --all

Download the text file:
[AutoEq setup on an Arm64 (Apple Silicon) M1 Mac v2.txt](https://github.com/jaakkopasanen/AutoEq/files/7043567/AutoEq.setup.on.an.Arm64.Apple.Silicon.M1.Mac.v2.txt)

**Comment by jaakkopasanen** [comment]
I updated AutoEq to use Python 3.10 and Tensorflow 2.9. This has proven to work for us at my company but since I don't have an M1 Mac I cannot test if this fixes everything in AutoEq. Could someone please test if AutoEq installs now without any extra tricks, @misty- perhaps?

**Comment by Pittvandewitt** [comment]
When installing the requirements it still has issues with `scipy` on macOS 11 specifically and with finding `tensorflow`.
Using `tensorflow-macos` instead of `tensorflow` gets us one step closer.
For scipy I had to find a whl file compatible with macOS 11: `pip install --pre -i https://pypi.anaconda.org/scipy-wheels-nightly/simple scipy`
After that things work as expected.

**Comment by jaakkopasanen** [comment]
Tensorflow part has been solved as Tensorflow is no longer a dependency.

Is scipy still an issue? I think some new wheels have been added to scipy recently.

**Comment by jaakkopasanen** [comment]
I upgraded soundfile to `~=0.11.0` and now everything installs correctly without any extra steps on my Mac M1 Pro :tada:
